File: backend/app/models/sim_request.py
import enum
from datetime import datetime
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Enum as SAEnum, Text, UniqueConstraint
from sqlalchemy.orm import relationship
from app.core.database import Base


class RequestStatus(str, enum.Enum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


class PermissionLevel(str, enum.Enum):
    VIEW = "view"    # can see card info only
    USE = "use"      # can send SMS / create tasks (implies view)


class SimAccessRequest(Base):
    """Records the approval workflow. Append-only once created; status is final."""
    __tablename__ = "sim_access_requests"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    modem_id = Column(Integer, ForeignKey("modems.id", ondelete="CASCADE"), nullable=False)
    status = Column(SAEnum(RequestStatus), default=RequestStatus.PENDING, nullable=False)
    requested_level = Column(SAEnum(PermissionLevel), default=PermissionLevel.USE, nullable=False)
    reason = Column(Text, nullable=True)
    admin_note = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    user = relationship("User", foreign_keys=[user_id])
    modem = relationship("Modem")


class SimGrant(Base):
    """当前有效授权，每个 (user, modem) 对唯一一行，审批通过或直接授权时 UPSERT。
    撤销时直接删除行；expires_at=None 表示永久有效。
    request_id=None 表示由管理员直接授权（非经申请流程）。
    """
    __tablename__ = "sim_grants"
    __table_args__ = (UniqueConstraint("user_id", "modem_id", name="uq_sim_grants_user_modem"),)

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    modem_id = Column(Integer, ForeignKey("modems.id", ondelete="CASCADE"), nullable=False)
    granted_level = Column(SAEnum(PermissionLevel), nullable=False)
    expires_at = Column(DateTime, nullable=True)      # None = 永久；到期后 get_user_modem_grants 忽略此行
    granted_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    request_id = Column(Integer, ForeignKey("sim_access_requests.id"), nullable=True)  # None = 直接授权
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    user = relationship("User", foreign_keys=[user_id])
    modem = relationship("Modem")
    granted_by = relationship("User", foreign_keys=[granted_by_id])

File: backend-go/models/sim_request.go
package models

import "time"

// 申请状态常量。
const (
	ReqPending  = "pending"  // 待审批
	ReqApproved = "approved" // 已批准
	ReqRejected = "rejected" // 已拒绝
)

// 权限级别常量：view 仅查看，use 可发送短信。
const (
	LevelView = "view"
	LevelUse  = "use"
)

// SimAccessRequest 记录用户对某 SIM 卡的访问申请，审批流程通过状态流转管理。
type SimAccessRequest struct {
	ID             uint      `gorm:"primaryKey" json:"id"`
	UserID         uint      `gorm:"not null" json:"user_id"`
	ModemID        uint      `gorm:"not null" json:"modem_id"`
	Status         string    `gorm:"size:16;not null;default:pending" json:"status"`      // pending/approved/rejected
	RequestedLevel string    `gorm:"size:16;not null;default:use" json:"requested_level"` // view 或 use
	Reason         *string   `gorm:"type:text" json:"reason"`                             // 申请理由（可选）
	AdminNote      *string   `gorm:"type:text" json:"admin_note"`                         // 审批备注（可选）
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

func (SimAccessRequest) TableName() string { return "sim_access_requests" }

// SimGrant 记录用户对某 SIM 卡的实际授权，(UserID, ModemID) 唯一。
// ExpiresAt=nil 表示永久授权；审批员角色通过 GetUserModemGrants 自动获得授权，无需此记录。
type SimGrant struct {
	ID           uint       `gorm:"primaryKey" json:"id"`
	UserID       uint       `gorm:"not null;uniqueIndex:uq_sim_grants_user_modem" json:"user_id"`
	ModemID      uint       `gorm:"not null;uniqueIndex:uq_sim_grants_user_modem" json:"modem_id"`
	GrantedLevel string     `gorm:"size:16;not null" json:"granted_level"` // view 或 use
	ExpiresAt    *time.Time `json:"expires_at"`                            // nil 表示永久
	GrantedByID  *uint      `json:"granted_by_id"`                         // 授权操作人
	RequestID    *uint      `json:"request_id"`                            // 关联的申请记录（直接授权时为 nil）
	CreatedAt    time.Time  `json:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at"`
}

func (SimGrant) TableName() string { return "sim_grants" }
